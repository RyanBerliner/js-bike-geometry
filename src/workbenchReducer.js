import {produce} from 'immer';

const SET_IMG_URL = 'set-image-url';
const SET_IMG_DETAILS = 'set-image-details';
const ADD_LAYER = 'add-layer';
const REMOVE_LAYER = 'remove-layer';
const UPDATE_LAYER = 'update-layer';
const SET_DRAWING_LAYER = 'set-drawing-layer';
const UPDATE_BRUSH_SETTINGS = 'update-brush-settings';
const UPDATE_STAGE_ZOOM = 'update-stage-zoom';

export const DISTORTION_ROTATIONAL = 'rotational';
export const DISTORTION_TRANSLATIONAL = 'translational';

export const INITIAL_DATA = {
  imageUrl: null,
  imageDetails: null,
  layerIds: [],
  layerData: {},
  drawingLayer: null,
  brushSettings: {
    fade: 50,
    size: 50,
    opacity: 100
  },
  stageZoom: 100
};

export const reducer = produce((draft, { type, payload }) => {
  switch(type) {
    case SET_IMG_URL:
      draft.imageUrl = payload;
      break;
    case SET_IMG_DETAILS:
      draft.imageDetails = payload === null ? null : {
        ...draft.imageDetails,
        ...payload
      };
      break;
    case ADD_LAYER:
      var id = (
        draft.layerIds.length === 0
          ? 1
          : parseInt(Math.max(...draft.layerIds)) + 1
      ).toString();

      draft.layerIds.push(id);
      draft.layerData[id] = payload;
      break;
    case REMOVE_LAYER:
      payload = payload.toString();
      draft.layerIds = draft.layerIds.filter(id => id !== payload);
      delete draft.layerData[payload];
      if (draft.drawingLayer === payload) draft.drawingLayer = null;
      break;
    case UPDATE_LAYER:
      draft.layerData[payload.id] = {
        ...draft.layerData[payload.id],
        ...payload.details,
      }
      break;
    case SET_DRAWING_LAYER:
      draft.drawingLayer = payload;
      break;
    case UPDATE_BRUSH_SETTINGS:
      draft.brushSettings[payload.setting] = payload.value;
      break;
    case UPDATE_STAGE_ZOOM:
      draft.stageZoom = payload;
      break;
    default:
      break;
  }

  return draft;
});

export const setImgUrl = url => ({ type: SET_IMG_URL, payload: url });
export const setImgDetails = details => ({ type: SET_IMG_DETAILS, payload: details });
export const addLayer = details => ({ type: ADD_LAYER, payload: details });
export const removeLayer = id => ({ type: REMOVE_LAYER, payload: id });
export const updateLayer = (id, details) => ({ type: UPDATE_LAYER, payload: {id, details} });
export const setDrawingLayer = id => ({ type: SET_DRAWING_LAYER, payload: id });
export const updateBrushSettings = (setting, value) => ({ type: UPDATE_BRUSH_SETTINGS, payload: {setting, value}});
export const updateStageZoom = stageZoom => ({ type: UPDATE_STAGE_ZOOM, payload: stageZoom });
