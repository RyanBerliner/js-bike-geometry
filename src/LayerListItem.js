import React from 'react';
import { DISTORTION_ROTATIONAL, DISTORTION_TRANSLATIONAL, setDrawingLayer, updateLayer } from './workbenchReducer';

export function LayerListItem({id, layerData, dispatch, isDrawing}) {
  const toggleDraw = () => {
    dispatch(setDrawingLayer(isDrawing ? null : id))
  }

  const updateLayerValue = event => {
    dispatch(updateLayer(id, {[event.target.name]: event.target.value}));
  }

  return <li className="py-3 px-2 border-bottom" id={`layer-${id}`}>
    <div className="d-xl-flex justify-content-between align-items-center mb-3">
      <div className="text-truncate lh-sm">
        <strong>{layerData.name}</strong><br />
        <span className="text-muted">{layerData.type}</span>
      </div>
      <div className="text-nowrap mt-xl-0 mt-2">
        <button className="btn ms-xl-2 me-2 btn-light" data-bs-target="#layer-modal" data-bs-toggle="modal" data-bs-layerid={id}>Edit</button>
        <button className={`btn ${isDrawing ? 'btn-danger' : 'btn-light'}`} onClick={toggleDraw}>
          <i className={`bi me-1 ${isDrawing ? 'opacity-75 bi-x-lg' : 'opacity-50 bi-brush-fill'}`}></i> {isDrawing ? 'Stop' : 'Draw'}
        </button>
      </div>
    </div>
    {layerData.type === DISTORTION_ROTATIONAL && <>
      <div className="row align-items-center">
        <label htmlFor={`rotationalangle-quick-${id}`} className="col-form-label col-3 text-nowrap lh-sm">Angle <br /><small className="text-muted">{layerData.rotationalangle} deg</small></label>
        <div className="col-9 d-flex align-items-center">
          <input type="range" className="form-range" id={`rotationalangle-quick-${id}`} name="rotationalangle" value={layerData.rotationalangle} onChange={updateLayerValue} min={-359} max={359} step={1} required={layerData.type === DISTORTION_ROTATIONAL} />
        </div>
      </div>
      <div className="row align-items-center">
        <label htmlFor={`rotationaloriginx-quick-${id}`} className="col-form-label col-3">OriginX</label>
        <div className="col-9 d-flex align-items-center">
          <input type="number" className="form-control form-control-sm" id={`rotationaloriginx-quick-${id}`} name="rotationaloriginx" value={layerData.rotationaloriginx} onChange={updateLayerValue} step={1} required={layerData.type === DISTORTION_ROTATIONAL} />
        </div>
      </div>
      <div className="row align-items-center">
        <label htmlFor={`rotationaloriginy-quick-${id}`} className="col-form-label col-3">OriginY</label>
        <div className="col-9 d-flex align-items-center">
          <input type="number" className="form-control form-control-sm" id={`rotationaloriginy-quick-${id}`} name="rotationaloriginy" value={layerData.rotationaloriginy} onChange={updateLayerValue} step={1} required={layerData.type === DISTORTION_ROTATIONAL} />
        </div>
      </div>
    </>}
    {layerData.type === DISTORTION_TRANSLATIONAL && <>
      <div className="row align-items-center">
        <label htmlFor={`translationalx-quick-${id}`} className="col-form-label col-3">TransX</label>
        <div className="col-9 d-flex align-items-center">
          <input type="number" className="form-control form-control-sm" id={`translationalx-quick-${id}`} name="translationalx" value={layerData.translationalx} onChange={updateLayerValue} step={1} required={layerData.type === DISTORTION_TRANSLATIONAL} />
        </div>
      </div>
      <div className="row align-items-center">
        <label htmlFor={`translationaly-quick-${id}`} className="col-form-label col-3">TransY</label>
        <div className="col-9 d-flex align-items-center">
          <input type="number" className="form-control form-control-sm" id={`translationaly-quick-${id}`} name="translationaly" value={layerData.translationaly} onChange={updateLayerValue} step={1} required={layerData.type === DISTORTION_TRANSLATIONAL} />
        </div>
      </div>
    </>}
  </li>
}