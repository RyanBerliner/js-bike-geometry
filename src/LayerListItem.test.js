import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {LayerListItem} from './LayerListItem';
import { updateLayer, setDrawingLayer } from './workbenchReducer';
import { DISTORTION_ROTATIONAL, DISTORTION_TRANSLATIONAL } from './workbenchReducer';

describe('layer list item', () => {
  describe('drawing control', () => {
    it('renders button properly when drawing and not drawing', () => {
      const props = {
        id: "1",
        layerData: {
          name: 'layer name',
          type: DISTORTION_TRANSLATIONAL,
          translationalx: 20,
          translationaly: 30,
        }
      };

      const { asFragment: notDrawing } = render(<LayerListItem isDrawing={false} {...props} />);
      const { asFragment: drawing } = render(<LayerListItem isDrawing={true} {...props} />);

      expect(notDrawing()).toMatchDiffSnapshot(drawing(), {contextLines: 0, stablePatchmarks: true});
    })

    it('properly dispatches change when toggled', async () => {
      const dispatch = jest.fn();

      const props = {
        id: "1",
        layerData: {
          name: 'layer name',
          type: DISTORTION_TRANSLATIONAL,
          translationalx: 20,
          translationaly: 30,
        },
        dispatch,
      }

      const {rerender} = render(<LayerListItem isDrawing={false} {...props} />)
      await userEvent.click(screen.getByText("Draw"))
      expect(dispatch.mock.calls[0][0]).toStrictEqual(setDrawingLayer("1"))

      rerender(<LayerListItem isDrawing={true} {...props} />)
      await userEvent.click(screen.getByText("Stop"))
      expect(dispatch.mock.calls[1][0]).toStrictEqual(setDrawingLayer(null))
    })
  })

  describe('rotational layer', () => {
    it('renders properly', () => {
      const { asFragment } = render(
        <LayerListItem
          id="1"
          layerData={{
            name: 'layer name',
            type: DISTORTION_ROTATIONAL,
            rotationalangle: 10,
            rotationaloriginx: 20,
            rotationaloriginy: 30,
          }}
        />
      );
  
      expect(asFragment()).toMatchSnapshot();
    })

    it('properly dispatches change of angle, originx, originy', () => {
      const dispatch = jest.fn();

      render(
        <LayerListItem
          id="1"
          layerData={{
            name: 'layer name',
            type: DISTORTION_ROTATIONAL,
            rotationalangle: 10,
            rotationaloriginx: 20,
            rotationaloriginy: 30,
          }}
          dispatch={dispatch}
        />
      )

      fireEvent.change(screen.getByRole('slider'), {target: {value: '15'}});
      expect(dispatch.mock.calls[0][0]).toStrictEqual(updateLayer('1', {rotationalangle: '15'}));

      fireEvent.change(screen.getByLabelText('OriginX'), {target: {value: '25'}});
      expect(dispatch.mock.calls[1][0]).toStrictEqual(updateLayer('1', {rotationaloriginx: '25'}));

      fireEvent.change(screen.getByLabelText('OriginY'), {target: {value: '35'}});
      expect(dispatch.mock.calls[2][0]).toStrictEqual(updateLayer('1', {rotationaloriginy: '35'}));
    })
  })

  describe('translational layer', () => {
    it('renders properly', () => {
      const { asFragment } = render(
        <LayerListItem
          id="1"
          layerData={{
            name: 'layer name',
            type: DISTORTION_TRANSLATIONAL,
            translationalx: 20,
            translationaly: 30,
          }}
        />
      );
  
      expect(asFragment()).toMatchSnapshot();
    })

    it('properly dispatches change of translate x, translate y', () => {
      const dispatch = jest.fn();

      render(
        <LayerListItem
          id="1"
          layerData={{
            name: 'layer name',
            type: DISTORTION_TRANSLATIONAL,
            translationalx: 20,
            translationaly: 30,
          }}
          dispatch={dispatch}
        />
      )

      fireEvent.change(screen.getByLabelText('TransX'), {target: {value: '25'}});
      expect(dispatch.mock.calls[0][0]).toStrictEqual(updateLayer('1', {translationalx: '25'}));

      fireEvent.change(screen.getByLabelText('TransY'), {target: {value: '35'}});
      expect(dispatch.mock.calls[1][0]).toStrictEqual(updateLayer('1', {translationaly: '35'}));
    })
  });
})