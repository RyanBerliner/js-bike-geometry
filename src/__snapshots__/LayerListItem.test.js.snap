// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layer list item drawing control renders button properly when drawing and not drawing 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ --- --- @@
-           class=\\"btn btn-light\\"
+           class=\\"btn btn-danger\\"
@@ --- --- @@
-             class=\\"bi me-1 opacity-50 bi-brush-fill\\"
+             class=\\"bi me-1 opacity-75 bi-x-lg\\"
@@ --- --- @@
-            Draw
+            Stop"
`;

exports[`layer list item rotational layer renders properly 1`] = `
<DocumentFragment>
  <li
    class="py-3 px-2 border-bottom"
    id="layer-1"
  >
    <div
      class="d-xl-flex justify-content-between align-items-center mb-3"
    >
      <div
        class="text-truncate lh-sm"
      >
        <strong>
          layer name
        </strong>
        <br />
        <span
          class="text-muted"
        >
          rotational
        </span>
      </div>
      <div
        class="text-nowrap mt-xl-0 mt-2"
      >
        <button
          class="btn ms-xl-2 me-2 btn-light"
          data-bs-layerid="1"
          data-bs-target="#layer-modal"
          data-bs-toggle="modal"
        >
          Edit
        </button>
        <button
          class="btn btn-light"
        >
          <i
            class="bi me-1 opacity-50 bi-brush-fill"
          />
           Draw
        </button>
      </div>
    </div>
    <div
      class="row align-items-center"
    >
      <label
        class="col-form-label col-3 text-nowrap lh-sm"
        for="rotationalangle-quick-1"
      >
        Angle 
        <br />
        <small
          class="text-muted"
        >
          10 deg
        </small>
      </label>
      <div
        class="col-9 d-flex align-items-center"
      >
        <input
          class="form-range"
          id="rotationalangle-quick-1"
          max="359"
          min="-359"
          name="rotationalangle"
          required=""
          step="1"
          type="range"
          value="10"
        />
      </div>
    </div>
    <div
      class="row align-items-center"
    >
      <label
        class="col-form-label col-3"
        for="rotationaloriginx-quick-1"
      >
        OriginX
      </label>
      <div
        class="col-9 d-flex align-items-center"
      >
        <input
          class="form-control form-control-sm"
          id="rotationaloriginx-quick-1"
          name="rotationaloriginx"
          required=""
          step="1"
          type="number"
          value="20"
        />
      </div>
    </div>
    <div
      class="row align-items-center"
    >
      <label
        class="col-form-label col-3"
        for="rotationaloriginy-quick-1"
      >
        OriginY
      </label>
      <div
        class="col-9 d-flex align-items-center"
      >
        <input
          class="form-control form-control-sm"
          id="rotationaloriginy-quick-1"
          name="rotationaloriginy"
          required=""
          step="1"
          type="number"
          value="30"
        />
      </div>
    </div>
  </li>
</DocumentFragment>
`;

exports[`layer list item translational layer renders properly 1`] = `
<DocumentFragment>
  <li
    class="py-3 px-2 border-bottom"
    id="layer-1"
  >
    <div
      class="d-xl-flex justify-content-between align-items-center mb-3"
    >
      <div
        class="text-truncate lh-sm"
      >
        <strong>
          layer name
        </strong>
        <br />
        <span
          class="text-muted"
        >
          translational
        </span>
      </div>
      <div
        class="text-nowrap mt-xl-0 mt-2"
      >
        <button
          class="btn ms-xl-2 me-2 btn-light"
          data-bs-layerid="1"
          data-bs-target="#layer-modal"
          data-bs-toggle="modal"
        >
          Edit
        </button>
        <button
          class="btn btn-light"
        >
          <i
            class="bi me-1 opacity-50 bi-brush-fill"
          />
           Draw
        </button>
      </div>
    </div>
    <div
      class="row align-items-center"
    >
      <label
        class="col-form-label col-3"
        for="translationalx-quick-1"
      >
        TransX
      </label>
      <div
        class="col-9 d-flex align-items-center"
      >
        <input
          class="form-control form-control-sm"
          id="translationalx-quick-1"
          name="translationalx"
          required=""
          step="1"
          type="number"
          value="20"
        />
      </div>
    </div>
    <div
      class="row align-items-center"
    >
      <label
        class="col-form-label col-3"
        for="translationaly-quick-1"
      >
        TransY
      </label>
      <div
        class="col-9 d-flex align-items-center"
      >
        <input
          class="form-control form-control-sm"
          id="translationaly-quick-1"
          name="translationaly"
          required=""
          step="1"
          type="number"
          value="30"
        />
      </div>
    </div>
  </li>
</DocumentFragment>
`;
